language/move-model/src/simplifier/mod.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use std::{fmt, str::FromStr};
+
+use crate::model::GlobalEnv;
+
+mod pass_inline;
+
+/// Available simplifications passes to run after tbe model is built
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum SimplificationPass {
+    Inline,
+}
+
+impl FromStr for SimplificationPass {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let r = match s {
+            "inline" => SimplificationPass::Inline,
+            _ => return Err(s.to_string()),
+        };
+        Ok(r)
+    }
+}
+
+impl fmt::Display for SimplificationPass {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Inline => write!(f, "inline"),
+        }
+    }
+}
+
+impl SimplificationPass {
+    pub fn run(&self, env: &mut GlobalEnv) -> Result<()> {
```

Comment:
Why does a simplification pass require a mut env? This is only available right after model building time, and we can't run simplification anywhere else then.

For the docgen use case of simplification, we would need both the original and the simplified spec block. At least that is how I thought about this. By default we show what the user has written, and then in an expandable block the simplified version.

Reply:
Good point. I wasn't making the distinctions clear in the code.

The code here is indeed only available right after model building time. This is for the case when I'd like to test that the simplification pass is working correctly (by proving that the spec coming out of the simplification still makes sense and verifies). For that, we need the ability to override the spec in the `GlobalEnv`, hence the `&mut` here.

For getting a simplified spec anywhere (including for the docgen), the API we are looking for is this:
`fn inline_all_exp_in_spec(env: &GlobalEnv, spec: Spec) -> Result<Spec> { .. }`, but I did not highlight this one but I should.

I'll improve the entrypoints and report back.

Reply:
Updated with a new commit (the last one):

[move-prover] better architecture for simplification pipeline
This creates a new trait, `SpecRewriter` that handles the
transformations on specs.

The `SpecRewriter` has three functions to be implemented by each
individual rewriter:
- `rewrite_module_spec`
- `rewrite_function_spec`
- `rewrite_inline_spec`

Multiple rewriters that implement the `SpecRewriter` trait can be
composed into a new rewriter that pipelines them. This logic is captured
in `SpecRewriterPipeline`, which also implements `SpecRewriter`.

A `SpecRewriter` has an option to write-back the newly generated spec
into a `&mut GlobalEnv` with the `override_with_rewrite` function. But
this is optional and only available in cases where a `&mut GlobalEnv`
can be obtained (e.g., during model building or right after the model
is built). If writing-back is not needed, the three `rewrite_*` funciton
should be sufficient.